### ISSUE TITLE
Subtle bug with NSObject.reactive.trigger(for:) in relation to URLSession

### ISSUE THREAD

**Issue by chrrasmussen** [ReactiveCocoa/ReactiveCocoa#3264]
# Problem description

I recently stumbled upon a subtle bug with `NSObject.reactive.trigger(for:)` (at least that is what it seems). I have an application that downloads some data from a JSON API using `URLSession.reactive.data(with:)`. It will download new data when navigating around in the application. After a while I noticed that the application stopped downloading more data, i.e. the signal producer from `URLSession.reactive.data(with:)` did not send "next" events anymore.

When the problem occurs, URLSession stops responding (does not call the callback function) to calls like `URLSession.dataTask()`, `URLSession.getAllTasks()`, etc. This also explains why `URLSession.reactive.data(with:)` does not send the "next" event.

After some digging I found out that the code started working again if I removed a call to `NSObject.reactive.trigger(for:)` (which was called in the same runloop iteration as `URLSession.reactive.data(with:)`). This function call seems irrelevant at first but somehow it affects the URLSession.

Other information that may be relevant:

- The problem occurs on both iOS 9 and iOS 10
- The problem occurs on both the simulator and the device
- The problem happens in normal usage, not only in tight loops (as seen in the example)
- I tried running all requests on the main thread but still got the problem

I havn't looked much into what can cause this behavior, as I opted for rewriting the code in question.

# Sample code

The minimal code needed to reproduce the problem is shown below. I have created an Xcode 8.1 project including ReactiveCococa 5.0.0-alpha.2 (using CococaPods, but the files are included in the repository for convenience): https://github.com/chrrasmussen/ReactiveCocoa-Trigger

```swift
import UIKit
import ReactiveSwift
import ReactiveCocoa

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        let config = URLSessionConfiguration.default
        let urlSession = URLSession(configuration: config)

        let request = URLRequest(url: Bundle.main.url(forResource: "Test", withExtension: "txt")!)

        for i in 1...200 {
            print("Requesting: #\(i)")
            let task = urlSession.reactive.data(with: request)

            task.startWithCompleted {
                print("Completed: #\(i)")
            }

            // This is the problematic code. Remove comment from line 26 to introduce bug.
            let viewController = UIViewController()
//            _ = viewController.reactive.trigger(for: #selector(UIViewController.viewDidAppear(_:)))
            // Observing the returned signal does not change anything. This one line is enough to break the URLSession.
        }
    }
}

```

Note: In addition to this code snippet, the code also requires access to an (empty) file named `Test.txt` which is expected to be in the main bundle. (See commit: https://github.com/chrrasmussen/ReactiveCocoa-Trigger/commit/10b9a2ec2bd49ee8f0117056664bee72034d28ac)

# What should happen?

The sample code will output `Requesting: #N` for each request (200 requests in the sample) with a corresponding `Completed: #N` log statement.

# What is actually happening?

The sample code will output `Requesting: #N` for each request but does not respond with a `Completed: #N` to all requests. At what point the code stops at is different every time, and sometimes it might even finish. If you don't see this behavior, please try again.

**Comment by andersio** [comment]
Yeah, something is definitely wrong in the Swift portion. ReactiveObjC does not have such problem. Investigating.

**Comment by andersio** [comment]
Fixed in #3265. Will push a bug fix release ASAP.

**Comment by chrrasmussen** [comment]
That's great! Thank you, @andersio :-)

**Comment by andersio** [comment]
Fixed in [5.0 alpha 3](https://github.com/ReactiveCocoa/ReactiveCocoa/releases/tag/5.0.0-alpha.3).
